code/traditional/TCA/TCA.py

```diff
@@ -87,50 +77,62 @@ def fit_predict(self, Xs, Ys, Xt1, Yt1):
         :param Yt: nt * 1, target label
         :return: Accuracy and predicted_labels on the target domain
         '''
-        Xs_new, Xt1_new = self.fit(Xs, Xt1)
-        clf = KNeighborsClassifier(n_neighbors=3)
-        
+        Xs_new, Xt_new = self.fit(Xs, Xt)
+        clf = KNeighborsClassifier(n_neighbors=1)
         clf.fit(Xs_new, Ys.ravel())
-        y_pred = clf.predict(Xt1_new)
-        acc = sklearn.metrics.accuracy_score(Yt1, y_pred)
-        
-        self.clf = clf
+        y_pred = clf.predict(Xt_new)
+        acc = sklearn.metrics.accuracy_score(Yt, y_pred)
+       
         return acc, y_pred
 
-    def fit_target(self, Xt2):
+    def fit_new(self, Xs, Xt, Xt2):
         '''
-        Map new Xt to the latent space create from self.fit using existing projection matrix
-        This Xt2 is different from Xt1 in the .fit
-        :param Xt : n_s, n_feature, new target feature
+        Map Xt2 to the latent space created from Xt and Xs
+        :param Xs : ns * n_feature, source feature
+        :param Xt : nt * n_feature, target feature
+        :param Xt2: n_s, n_feature, target feature to be mapped
         '''
-        if not np.all(self.W_proj):
-            raise AssertionError('No projection matrix found, use .fit to find new feature of source and target')
+        # Computing projection matrix A from Xs an Xt
+        X = np.hstack((Xs.T, Xt.T))
+        X /= np.linalg.norm(X, axis=0)
+        m, n = X.shape
+        ns, nt = len(Xs), len(Xt)
+        e = np.vstack((1 / ns * np.ones((ns, 1)), -1 / nt * np.ones((nt, 1))))
+        M = e * e.T
+        M = M / np.linalg.norm(M, 'fro')
+        H = np.eye(n) - 1 / n * np.ones((n, n))
+        K = kernel(self.kernel_type, X, None, gamma=self.gamma)
+        n_eye = m if self.kernel_type == 'primal' else n
+        a, b = np.linalg.multi_dot([K, M, K.T]) + self.lamb * np.eye(n_eye), np.linalg.multi_dot([K, H, K.T])
+        w, V = scipy.linalg.eig(a, b)
+        ind = np.argsort(w)
+        A = V[:, ind[:self.dim]]
         
-        # Reshape to make it consistent with other methods
+        # Compute kernel with Xt2 as target and X as source
         Xt2 = Xt2.T
-        
-        # Compute kernel with respect to self.X_new_src
-        K = kernel(self.kernel_type, X1 = Xt2, X2 = self.X_new_src, gamma=self.gamma)
+        K = kernel(self.kernel_type, X1 = Xt2, X2 = X, gamma=self.gamma)
         
         # New target features
-        Xt2_new = K @ self.W_proj
+        Xt2_new = K @ A
         
         return Xt2_new
     
-    def fit_predict_target(self, Xt2, Yt2):
+    def fit_new_predict(self, Xt, Xs, Ys, Xt2, Yt2):
         '''
```

Comment:
I think it got quite messy here, so these are parameters requires by `fit_new_predict` :
- `Xs` and `Xt`: to be passed into `fit` to create `Xs_new` (ignore `Xt_new`) then these two are also passed into `fit_new` to create `Xt2_new`
- `Ys`: to be paired with `Xs_new` to fit clf
- `Xt2 and Yt2`:  to create `Xt2_new` and also to evaluate on clf

Also I got trouble finding a good name for `Xt2` is it okay enough, any suggestion?
To be consistent with previous code, all variables followed by `new` are transformed feature by TCA.
Sorry to update this on weekend, it's okay you can reply me on weekday or anytime that suits you.